### ISSUE TITLE
[R-Forge #1957] Optional Arguments for Printing

### ISSUE THREAD

**Issue by arunsrinivasan** [Rdatatable/data.table#645]
Submitted by: Yike Lu; Assigned to: Nobody; [R-Forge link](https://r-forge.r-project.org/tracker/index.php?func=detail&aid=1957&group_id=240&atid=978)

It would be nice to have a few simple optional arguments for displaying/printing data.tables, for example:
1) Turn off row numbers,
2) Turn off smart table wrapping
3) Force print all entries
4) Put spaces between "by" groupings for higher groupings... e.g. DT[, sum(x), by=list(a,b,c)]  prints as

a    b     c   V1
1    1     1
1    1     2

1    0     3

2    1     1
2    1     2
2    1     3

Perhaps a depth variable can be specified here.

5) underline/underscores for column header, horizontal bar to separate key and nonkey columns.

**Comment by MichaelChirico** [comment]
1) & 3) are done, AFAIK.

Not sure anything can be done about 2), if I understand correctly?

4) Could go for some more formalization, not sure exactly what was intended.

First part of 5) seems pretty simple, maybe a `name.sep` or `bezel` argument; second part seems a bit tougher.

As mentioned in #1483 fulfilling this should probably entail creating an `.Rd` for `print.data.table`. And exporting `print.data.table`?

**Comment by franknarf1** [comment]
Personally, 2 and 4 would be very handy. I have a wrapper to do 4 (though the BY vars still show up in my "blank" rows) and just expand my R console across more screens for 2. 

(I do this because I have been reviewing a feed of data periodically, summarized to about 20 groups, each having about four rows for subgroup.)

**Comment by MichaelChirico** [comment]
@franknarf1 could you add some detail about what 4) means for you? I think I'm missing something in the formatting from R-Forge.

From how I read it, it's something like

    DT <- data.table(a = rep(1:2, each = 3), b = rep(1:3, 2))

    DT[ , 1L, by = .(a, b)]
```
DT[ , 1L, by = .(a, b)]
#    a b V1
# 1: 1 1  1
# 2: 1 2  1
# 3: 1 3  1
#                   # <- blank row here as we switch 'a' group from 1 to 2
# 4: 2 1  1
# 5: 2 2  1
# 6: 2 3  1
```

And the `depth` Yike had in mind was saying how frequently we use the spacing for more fine grouping -- if we did `.(a, b, c)` with `depth = 2`, we'd put spaces after every switch of `b`.

Does that sound like what you have in mind?

**Comment by arunsrinivasan** [comment]
@MichaelChirico would it be possible for you to wrap all print method related features/enhancements together (from multiple issues) on to one issue (with a check box for each). It'd be nice to also have the issue number and the person who filed the report there so as to credit accordingly. It'd be easier to lookup/update.

**Comment by MichaelChirico** [comment]
@arunsrinivasan done at #1523

**Comment by franknarf1** [comment]
> could you add some detail about what 4) means for you? I think I'm missing something in the formatting from R-Forge. [...] From how I read it, it's something like [...]

@MichaelChirico Sure, yeah your reading is the same as mine. 

Here's what I do now (somewhat simplified):

    DT[, {
        subdat  = lapply(.SD, as.character)
        newrow0 = rep(list(""), length(subdat))
        rbind(
          .SD, 
          newrow0
        )
    }, by=a]

I've realized that `DT[, {print(.SD); cat("\n")}, by=a]` is a good alternative, though it does print the column names on every `.SD`. An option `col.names=FALSE` would solve that.

I only noticed Yike's `depth` idea when you pointed it out now. I think you have the right interpretation for it, but I have no need for it myself right now (since I only browse against a single grouping variable).

---- 

In case it's helpful in understanding what I'm after, here is my unsimplified use-case. Really, `print.data.table` is fine as-is; but I just wanted to chime in and say that yes, inserting blank lines is a thing that I actually do, too.

    DT2 = setDT(structure(list(event = c("A", "A", "A", "A", "B", "B", "B", "B", 
    "C", "C", "C", "C"), v1 = c("x", "x", "x", "x", "x", "x", "x", 
    "x", "x", "x", "x", "x"), v2 = c("xxxxxxxxxxx", "xxxxxxxxxxx", 
    "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", 
    "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx"
    ), v3 = c("xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", 
    "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx", 
    "xxxxxxxxxxx", "xxxxxxxxxxx", "xxxxxxxxxxx"), v4 = c("xxx", "xxx", 
    "xxx", "xxx", "xxx", "xxx", "xxx", "xxx", "xxx", "xxxx", "xxx", 
    "xxx"), v5 = c("xxxxxxxxx", "xxxxx", "xxxxx", "xx", "xxxxxxxxxxxxxxxxxxxxxx", 
    "xxxxxxxx", "xxxxx", "xxxxx", "xxxxxxxxxxxxxxxxxxxxxx", "xxxxxxxxxxxxxxxxx", 
    "xxxxxxxxxxxxxxxxxxxxxxx", "xxxxxxxxxxxxxxxxx"), v6 = c("xx", 
    "xx", "xx", "xx", "xxx", "xx", "xx", "xx", "xxx", "xx", "xx", 
    "xx")), row.names = c(NA, -12L), class = "data.frame", .Names = c("event", 
    "v1", "v2", "v3", "v4", "v5", "v6")))

    eventDT = data.table(event = LETTERS[1:3], a1 = c(100,200,300), a2 = c("aaa","bbb","ccc"))

    print(DT2[,{
        subdat = .SD
        newrow0 = lapply(.SD, function(x) as("",class(x)))
        newrow  = newrow0
        newrow["v5"] = eventDT[.BY, paste(a1, a2), on="event"]
        rbind(
          .SD, 
          newrow0,
          newrow,
          newrow0
        )
    }, by=event], row.names=FALSE)

... which gives ...

     event v1          v2          v3   v4                      v5  v6
         A  x xxxxxxxxxxx xxxxxxxxxxx  xxx               xxxxxxxxx  xx
         A  x xxxxxxxxxxx xxxxxxxxxxx  xxx                   xxxxx  xx
         A  x xxxxxxxxxxx xxxxxxxxxxx  xxx                   xxxxx  xx
         A  x xxxxxxxxxxx xxxxxxxxxxx  xxx                      xx  xx
         A                                                            
         A                                                 100 aaa    
         A                                                            
         B  x xxxxxxxxxxx xxxxxxxxxxx  xxx  xxxxxxxxxxxxxxxxxxxxxx xxx
         B  x xxxxxxxxxxx xxxxxxxxxxx  xxx                xxxxxxxx  xx
         B  x xxxxxxxxxxx xxxxxxxxxxx  xxx                   xxxxx  xx
         B  x xxxxxxxxxxx xxxxxxxxxxx  xxx                   xxxxx  xx
         B                                                            
         B                                                 200 bbb    
         B                                                            
         C  x xxxxxxxxxxx xxxxxxxxxxx  xxx  xxxxxxxxxxxxxxxxxxxxxx xxx
         C  x xxxxxxxxxxx xxxxxxxxxxx xxxx       xxxxxxxxxxxxxxxxx  xx
         C  x xxxxxxxxxxx xxxxxxxxxxx  xxx xxxxxxxxxxxxxxxxxxxxxxx  xx
         C  x xxxxxxxxxxx xxxxxxxxxxx  xxx       xxxxxxxxxxxxxxxxx  xx
         C                                                            
         C                                                 300 ccc    
         C                                                            
     event v1          v2          v3   v4                      v5  v6

**Comment by MichaelChirico** [comment]
@franknarf1 i'm not following your example here. the point is just to show how to force the alignment to be consistent?

**Comment by MichaelChirico** [comment]
the other unresolved question I have on this feature's implementation is how it should interact with `topn` and `nrows`. Should those cascade down by group, or should they apply to the total output? If there are many groups, this could create massive output.

**Comment by franknarf1** [comment]
@MichaelChirico 

> @franknarf1 i'm not following your example here. the point is just to show how to force the alignment to be consistent?

Yes, it's pretty much the same as `lapply(split(DT, by="g"), print, ...)` except columns stay aligned. 

So if print had a column-widths argument, it would solve my problem and also probably have wider applicability than this gaps-between-groups idea. Back to the example above, I would write a wrapper to do...

```
# note: this is non-working code, just showing what desired behavior would be

s = split(DT2, by="event", keep.by=FALSE)
wids = Reduce(pmax, lapply(s, function(x) sapply(x, function(v) max(nchar(v)))))
for (i in seq_along(s)){
  cat(sprintf("\nGRP: %s\n", names(s)[i]))
  print(s[[i]], width = wids + 1)
}

# GRP: A
# v1 v2           v3           v4    v5                       v6
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxxxxxx                xx  
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxx                    xx  
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxx                    xx  
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xx                       xx  
# 
# GRP: B
# v1 v2           v3           v4    v5                       v6
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxxxxxxxxxxxxxxxxxxx   xxx 
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxxxxx                 xx  
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxx                    xx  
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxx                    xx  
# 
# GRP: C
# v1 v2           v3           v4    v5                       v6
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxxxxxxxxxxxxxxxxxxx   xxx 
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxxx  xxxxxxxxxxxxxxxxx        xx  
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxxxxxxxxxxxxxxxxxxxx  xx  
# x  xxxxxxxxxxx  xxxxxxxxxxx  xxx   xxxxxxxxxxxxxxxxx        xx  
```

("width" is what the argument is called in `write.fwf`.)

This would be a lot simpler than my weird `rbind` above, and would give nicer output.

> the other unresolved question I have on this feature's implementation is how it should interact with `topn` and `nrows`. Should those cascade down by group, or should they apply to the total output? If there are many groups, this could create massive output.

(Assuming you pursue this instead of widths...) I think it would make more sense for them to cascade (similar to the `...` in the `lapply` I just mentioned), since I don't know what it would mean to apply them to the table as a whole. Regarding the massive output, for me this approach to printing is just for manual browsing. I send the output to a text file and browse that (rather than looking at it in the console). fwiw.

----

By the way, I came across [a related idea on SO this week](https://stackoverflow.com/questions/45621385/convert-group-variable-to-group-name-after-dplyrgroup-by-in-r):  when a table is keyed, write the first value of each group only once. The OP mentioned that pandas supports this, and I found [the docs](https://pandas.pydata.org/pandas-docs/stable/groupby.html#groupby-with-multiindex). An example from there:

```
first  second  third
bar    doo     one     -1.131345
               two     -0.089329
baz    bee     one      0.337863
               two     -0.945867
foo    bop     one     -0.932132
               two      1.956030
qux    bop     one      0.017587
               two     -0.016692
```

Also [a blog post](http://bconnelly.net/2013/10/summarizing-data-in-python-with-pandas/).
